Modules/IO/DCMTK/src/itkDCMTKImageIO.cxx

```diff
@@ -28,25 +28,108 @@
+ * (that does not have preamble).
+ * @param file A stream to test if the file is dicom like
+ * @return true if the structure of the file is dicom like
+ */
+bool
+isNoPreambleDicom(std::ifstream & file) // NOTE: Similar function is in itkGDCMImageIO.cxx
 {
-  char preamble[132] = "";
+  file.seekg(0, std::ios::beg);
 
-  file.read(preamble, sizeof(preamble));
+  // Adapted from https://stackoverflow.com/questions/2381983/c-how-to-read-parts-of-a-file-dicom
+  /* This heuristic tries to determine if the file follows the basic structure of a dicom file organization.
+   * Any file that begins with the a byte sequence
+   * where groupNo matches below will be then read several SOP Instance sections.
+   */
 
-  return (preamble[128] == 'D' && preamble[129] == 'I' && preamble[130] == 'C' && preamble[131] == 'M');
+  unsigned short groupNo = 0xFFFF;
+  unsigned short tagElementNo = 0xFFFF;
+  do
+  {
+    file.read(reinterpret_cast<char *>(&groupNo), sizeof(unsigned short));
+    itk::ByteSwapper<unsigned short>::SwapFromSystemToLittleEndian(&groupNo);
+    file.read(reinterpret_cast<char *>(&tagElementNo), sizeof(unsigned short));
+    itk::ByteSwapper<unsigned short>::SwapFromSystemToLittleEndian(&tagElementNo);
+
+    if (groupNo != 0x0002 && groupNo != 0x0008) // Only groupNo 2 & 8 are supported without preambles
+    {
+      return false;
+    }
+
+    char vrcode[3] = { '\0', '\0', '\0' };
+    file.read(vrcode, 2);
+
+    long              length = std::numeric_limits<long>::max();
+    const std::string vr{ vrcode };
+    if (vr == "AE" || vr == "AS" || vr == "AT" || vr == "CS" || vr == "DA" || vr == "DS" || vr == "DT" || vr == "FL" ||
+        vr == "FD" || vr == "IS" || vr == "LO" || vr == "PN" || vr == "SH" || vr == "SL" || vr == "SS" || vr == "ST" ||
+        vr == "TM" || vr == "UI" || vr == "UL" || vr == "US")
+    {
+      // Explicit VR (value representation stored in the file)
+      unsigned short uslength = 0;
+      file.read(reinterpret_cast<char *>(&uslength), sizeof(unsigned short));
+      itk::ByteSwapper<unsigned short>::SwapFromSystemToLittleEndian(&uslength);
+      length = uslength;
+    }
+    else
+    {
+      // Implicit VR (value representation not stored in the file)
+      char lengthChars[4] = { vrcode[0], vrcode[1], '\0', '\0' };
+      file.read(lengthChars + 2, 2);
+
+      auto * uilength = reinterpret_cast<unsigned int *>(lengthChars);
+      itk::ByteSwapper<unsigned int>::SwapFromSystemToLittleEndian(uilength);
+
+      length = (*uilength);
+    }
+    if (length <= 0)
+    {
+      return false;
+    }
+    file.ignore(length);
+    if (file.eof())
+    {
+      return false;
+    }
+  } while (groupNo == 2);
+
+#if defined(NDEBUG)
+  std::ostringstream itkmsg;
+  itkmsg << "No DICOM magic number found, but the file appears to be DICOM without a preamble.\n"
+         << "Proceeding without caution.";
+  itk::OutputWindowDisplayDebugText(itkmsg.str().c_str());
+#endif
```

Comment:
Is this extra debug output really necessary? If so, is it really meant to be `#if defined(NDEBUG)`?  Or maybe `#ifndef NDEBUG`? (NDEBUG = _not_ debug)

Reply:
Good catch. I agree, it should be `#ifndef NDEBUG` as everywhere else. It was probably just a typo back then when it was implemented (I just moved the code to a new location within the file).

Reply:
`Proceeding without caution` comment does not make sense either. The DICOM tag detection mechanism is pretty safe there and the test is anyway just a quick test to check if the file looks like DICOM (if the file is actually not valid DICOM then reading will fail).

Reply:
Thanks, glad I could be of help 😃 Honestly I always found the name of the macro `NDEBUG` quite confusing myself. I guess we'd have to ask the C standard committee for a better name  😸

Reply:
I guess by the time they wanted to add that feature, `DEBUG` macro was quite widely used, with perhaps different meaning, and they didn't want to break all that, so they came up with `NDEBUG`, which nobody used 😄

Reply:
BTW The use of `std::ostringstream` appears unnecessary in this case. I would rather pass the literal string to OutputWindowDisplayDebugText directly:

```cpp
  itk::OutputWindowDisplayDebugText(
    "No DICOM magic number found, but the file appears to be DICOM without a preamble.\n"
    "Proceeding without caution.");
```

Reply:
@lassoan Thanks for using `itkDebugMacro` instead  👍